feat(prefect-server): make Deployment strategy configurable

stigok opened the pull request:
### Summary

Adds value `server.deploymentStrategyType`.

Useful when using sqlite database backend on a mounted volume, where the volume can only be mounted on a single node at a time. If the strategy is "OrderedReady" and the new pod is scheduled on a different node, the deployment will hang indefinitely.

### Requirements

- [x] [Contributing guide](https://github.com/PrefectHQ/prefect-helm/?tab=readme-ov-file#contributing) has been read
- [x] Title follows the [conventional commits](https://www.conventionalcommits.org) format
- [ ] Body includes `Closes <issue>`, if available
- [x] Added/modified configuration includes a descriptive comment for documentation generation
- [x] Unit tests are added/updated
- [ ] Deprecation/removal checks are added in `templates/NOTES.txt`
- [x] Relevant labels are added
- [x] `Draft` status is used until ready for review

stigok committed:
feat: add value for deployment.spec.strategy

Adds value `server.updateStrategy`.

Useful when using sqlite database backend on a mounted volume, where the
volume can only be mounted on a single node at a time. If the strategy
type is `RollingUpdate` and the new pod is scheduled on a different node,
the deployment will hang indefinitely, as the new pod is not able to
start as long as the disk it requires is mounted to the old one.

stigok committed:
feat: add value for deployment.spec.strategy.type

Adds value `server.deploymentStrategyType`.

Useful when using sqlite database backend on a mounted volume, where the
volume can only be mounted on a single node at a time. If the strategy
is "OrderedReady" and the new pod is scheduled on a different node, the
deployment will hang indefinitely.

mitchnielsen reviewed (changes requested):
Thanks for this @stigok! Couple small suggestions.

stigok commented on charts/prefect-server/templates/server-deployment.yaml:
```diff
@@ -18,6 +18,10 @@ spec:
   selector:
     matchLabels: {{- include "common.labels.matchLabels" . | nindent 6 }}
       app.kubernetes.io/component: server
+  {{- if not (empty .Values.server.deploymentStrategyType) }}
+  strategy:
+    type: {{ .Values.server.deploymentStrategyType }}
+  {{- end }}
```
I thought it'd be nice to piggy-back on the Kubernetes default instead, but seems the most used practice here is to set explicit defaults. Guess it helps having defaults in the documentation, too.

stigok commented:
- Removed a bogus test
- Renamed the title of the remaining test to match existing format
- Updated code with received suggestions

I should probably not have force-pushed until you re-reviewed. I'll remember for next time.

I am now considering if a better road is to expose `server.deploymentStrategy` instead, so that we are preparing to allow setting the `Deployment.spec.strategy.rollingUpdate` as well as just `Deployment.spec.strategy.type`. If this is to be done later, the current state of this PR would probably require a rename/deprecation of the `deploymentStrategyType` to be performed. What do you think?

mitchnielsen commented:
> I am now considering if a better road is to expose server.deploymentStrategy instead, so that we are preparing to allow setting the Deployment.spec.strategy.rollingUpdate as well as just Deployment.spec.strategy.type. If this is to be done later, the current state of this PR would probably require a rename/deprecation of the deploymentStrategyType to be performed. What do you think?

Fair point, maybe that's worth doing now instead of having to deprecate and remove it later.

It made me think of even going one step further to expose `server.deployment.strategy`, because I see a pattern in many charts for the top-level key to match the template for the Kubernetes object it'll create. Looks like we don't really follow that pattern here, so sticking with `server.deploymentStrategy` is probably fine.

stigok committed:
update values schema with new additions

stigok commented:
Alright, should be set for a new review. Just call for a squash or rebase if you want. I also don't mind if you make edits yourself.

stigok committed:
use string values in examples to match schema

stigok commented on charts/prefect-server/values.schema.json:
```diff
@@ -200,6 +200,35 @@
+              "description": "Rolling update config params. Should only be specified if updateStrategy is set to \"RollingUpdate\".",
+              "properties": {
+                "maxSurge": {
+                  "type": "string",
```
A `string` type will make it "legal" to both specify `"1"` and `"25%"`. Isn't that the most fitting type here?

mitchnielsen replied:
Agreed, c3558141d96e9b575e4bb4fcf578fe1e43619966 looks good to me.

mitchnielsen commented on charts/prefect-server/templates/server-deployment.yaml:
```diff
@@ -18,6 +18,7 @@ spec:
   selector:
     matchLabels: {{- include "common.labels.matchLabels" . | nindent 6 }}
       app.kubernetes.io/component: server
+  strategy: {{ toYaml .Values.controller.updateStrategy | nindent 4 }}
```
Ah, this should be `server` instead of `controller` - here's a diff with the fix and a test that will catch that:

```diff
diff --git a/charts/prefect-server/templates/server-deployment.yaml b/charts/prefect-server/templates/server-deployment.yaml
index 170b9bb..522cabe 100644
--- a/charts/prefect-server/templates/server-deployment.yaml
+++ b/charts/prefect-server/templates/server-deployment.yaml
@@ -18,7 +18,7 @@ spec:
   selector:
     matchLabels: {{- include "common.labels.matchLabels" . | nindent 6 }}
       app.kubernetes.io/component: server
-  strategy: {{ toYaml .Values.controller.updateStrategy | nindent 4 }}
+  strategy: {{ toYaml .Values.server.updateStrategy | nindent 4 }}
   template:
     metadata:
       {{- if .Values.server.podAnnotations }}
diff --git a/charts/prefect-server/tests/server_test.yaml b/charts/prefect-server/tests/server_test.yaml
index d25b6db..30839f3 100644
--- a/charts/prefect-server/tests/server_test.yaml
+++ b/charts/prefect-server/tests/server_test.yaml
@@ -601,6 +602,28 @@ tests:
           path: .spec.strategy.type
           value: Recreate
 
+  - it: Should set RollingUpdate strategy fields
+    set:
+      server:
+        updateStrategy:
+          type: RollingUpdate
+          rollingUpdate:
+            maxSurge: "1"
+            maxUnavailable: "25%"
+    asserts:
+      - template: server-deployment.yaml
+        equal:
+          path: .spec.strategy.type
+          value: RollingUpdate
+      - template: server-deployment.yaml
+        equal:
+          path: .spec.strategy.rollingUpdate.maxSurge
+          value: "1"
+      - template: server-deployment.yaml
+        equal:
+          path: .spec.strategy.rollingUpdate.maxUnavailable
+          value: "25%"
+
   - it: Should set autoscaling configuration when enabled
     set:
       server:
```

stigok replied:
Good catch and good additional test.

stigok committed:
fix wrong variable value reference

stigok committed:
add test for .spec.strategy.rollingUpdate

mitchnielsen committed:
Merge branch 'main' into main
